### PR TITLE
Add an option to not build the private part of the Standard Library

### PR DESCRIPTION
**Pull request by edymtt**
This can be useful to reduce build times in certain configurations (e.g.
when we don't run tests).

Addresses rdar://86474460

### CODE REVIEW

**Comment by edymtt** [comment]
@swift-ci please python lint

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build
@swift-ci please test with toolchain and preset

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build
@swift-ci please test with toolchain and preset

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
Hey @edymtt, I think I'm pursuing a similar change for a similar reason, see <https://github.com/apple/swift/pull/40211/files>, and in my case the "trigger" is just setting SWIFT_BUILD_TEST_SUPPORT_MODULES to Off. Not sure which approach is better, but I wonder if you could try my PR and see if it helps in your situation too?

**Comment by edymtt** [comment]
I think our intents are different -- I am interested in having a way to build the Darwin overlay without the private part of the Standard Library, at the moment those are gated on the same flags (`SWIFT_BUILD_SDK_OVERLAY OR SWIFT_BUILD_TEST_SUPPORT_MODULES`), so at the moment I'm proposing adding a new flag to allow this (but I'm not entirely sure this is the correct approach).

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
Are you building with SWIFT_BUILD_SDK_OVERLAY On?

**Comment by edymtt** [comment]
Correct, `stdlib_S_standalone_darwin_x86_64,build` does not set `build-swift-dynamic-sdk-overlay` to 0.

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build
@swift-ci please test with toolchain and preset

**Review by gottesmm — APPROVED** [review]
If the tests pass, LGTM.

File: stdlib/CMakeLists.txt [hunk 1]
```diff
@@ -164,6 +164,10 @@ option(SWIFT_ENABLE_REFLECTION
   "Build stdlib with support for runtime reflection and mirrors."
   TRUE)
 
+option(SWIFT_BUILD_PRIVATE_STDLIB
```
Comment (compnerd) [inline stdlib/CMakeLists.txt:167]: I slightly think that `SWIFT_STDLIB_BUILD_PRIVATE` might be a better spelling for the option, but 🤷‍♂️ .

**Review by compnerd — APPROVED** [review]
I don't have a concern with the addition of the option, but perhaps we can find a better spelling of the option.

File: stdlib/CMakeLists.txt (hunk 1 above)
Comment (edymtt) [reply stdlib/CMakeLists.txt]: Good point -- tackled in d39805b3ff603657923f344b92f8fe676915a522

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build
@swift-ci please test with toolchain and preset

**Comment by edymtt** [comment]
@swift-ci please python lint

**Comment by edymtt** [comment]
@swift-ci please smoke test

**Comment by edymtt** [comment]
@swift-ci please test Apple Silicon

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
Using stdlib_S_standalone_darwin_x86_64,build `stdlib/private` is not built

https://ci.swift.org/job/swift-PR-stdlib-with-toolchain-osx-preset/86/

**Comment by edymtt** [comment]
preset=stdlib_S_standalone_darwin_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
> This failure for stdlib_S_standalone_darwin_x86_64,build,test is expected, since the toolchain we use is not new enough

That's right, feel free to ignore it, and merge anyway (if other tests pass).

**Review by kubamracek — APPROVED** [review]

**Comment by edymtt** [comment]
The macOS toolchain issue is the same as https://ci.swift.org/view/Packages/job/oss-swift-package-macos/6424/
and it's addressed in https://github.com/apple/swift-package-manager/pull/3952

**Pull request merged**